### PR TITLE
Demonstrating errors in code generation

### PR DESCRIPTION
**Pull request by nickerso**
I have added code generation for my "go to" test model (sine approximations with imports) to demonstrate an issue that I am seeing in the code generation. There is some evidence that this is related to #597, but as well as the ordering being incorrect, the `VARIABLE_INFO` is getting the name of the variables wrong. While the variable info correctly identifies the "source" component, it is using the name of the variable from the top-level component in the encapsulation and import hierarchy.

You can see these differences in specifically in https://github.com/nickerso/libcellml/commit/41db32b632b909bfea0fba5a04b114da841f8770. This is a problem in both the C and Python profiles.

### CODE REVIEW

**Comment by agarny** [comment]
Good catch! FWIW, I would indeed expect the behaviour to be the same indepdent on the profile chosen. Anyway, it should be an easy fix. I will do it tomorrow once PRs #596 and #599 have been merged in.

**Comment by nickerso** [comment]
> once PRs #596 and #599 have been merged in

Surely the fix can be done without a merge...this is completely blocking me at present...I'm building my own libcellml so not worried about changes being merged in...pretty please :)

**Comment by agarny** [comment]
I think the problem (besides issue #597) is not with the generator itself, but rather with the flattening.

**Comment by agarny** [comment]
> I think the problem (besides issue #597) is not with the generator itself, but rather with the flattening.

I take it back. It is indeed an issue with the generator. Argh!

**Comment by nickerso** [comment]
> I take it back. It is indeed an issue with the generator.

phew! was pretty sure I checked the flattened model and it looked good. This issue came up as I try to navigate through the collection of equivalent variables to find the one that is used in the generated code and I can never get a match as the component and variable names are out of sync.

**Comment by agarny** [comment]
> > I take it back. It is indeed an issue with the generator.
> 
> phew! was pretty sure I checked the flattened model and it looked good. This issue came up as I try to navigate through the collection of equivalent variables to find the one that is used in the generated code and I can never get a match as the component and variable names are out of sync.

Yes, I printed the flattened model and that's where I realised my mistake. :) The problem is buried in our generator, but I am slowly getting there, so you will have a fix before the end of the day. Definitely.

**Comment by agarny** [comment]
Ok, I am done with this. Happy for you, @nickerso, @hsorby and @kerimoyle, to review and merge this in.

**Comment by agarny** [comment]
@nickerso, since you can't review it as such. I am happy to "review" it on your behalf...

**Comment by nickerso** [comment]
Mild panic as this didn't fix my issue, but then realised that I have the ordering fix from #597 also merged into my local source. So ignoring that, yes, this fixes the bug in the variable names and I would happily approve this if I hadn't started it :)

**Comment by agarny** [comment]
Eh, it's not the time for mild panic! :) Anyway, I am going to approve it on your behalf and leave it to @hsorby or @kerimoyle to confirm.

**Review by agarny — APPROVED** [review]
This is a brilliant PR. I am all in favour of it. No reservations whatsoever! (If only all PRs could be of that quality. One can only dream...)

File: src/generator.cpp
Comment (kerimoyle) [inline src/generator.cpp:156]: A variable that is reset will have an initial value too? Neither a constant nor a state?

**Review by kerimoyle — COMMENTED** [review]
Which is the bit that fixes the bug? There are lots of changes that seem a bit arbitrary?

File: tests/resources/deriv_approx_sin.xml
Comment (kerimoyle) [inline tests/resources/deriv_approx_sin.xml:6]: This is now testing something different ... even if we are phasing out the initialisation with variable thing, we should keep the tests?

File: src/generator.cpp
Comment (agarny) [reply src/generator.cpp:156]: We are not yet generating code for resets, so we will see what needs to be done when we do.

**Comment by agarny** [comment]
> Which is the bit that fixes the bug? There are lots of changes that seem a bit arbitrary?

This PR contains some code clean up (as a result of working on this PR), but otherwise there are a couple of key places that fix the bug. I am going to annotate them.

File: tests/resources/deriv_approx_sin.xml
Comment (agarny) [reply tests/resources/deriv_approx_sin.xml:6]: This was @nickerso's change and I think it's fine since we are not currently handling the initialisation of a variable from another variable.

File: src/generator.cpp
Comment (agarny) [inline src/generator.cpp:527]: 1) This is the change that directly fixes @nickerso's issue. Before we were only updating the component information, but not the variable information which we now do through our call to `setVariable()`. Note, however, that we pass it `false` since we don't want to check for the initial value.

File: src/generator.cpp
Comment (agarny) [inline src/generator.cpp:161]: 2) Our new `setVariable()` method now also updates `mVariable`, but also keeps track of the variable object that contains the initial value of a variable (`mInitialValueVariable`).

File: src/api/libcellml/generator.h
Comment (agarny) [inline src/api/libcellml/generator.h:81]: 3) Because of points 1 and 2, the `component()` is always going to be the parent component of `variable()`, which makes our API better than before.

File: src/generator.cpp
Comment (agarny) [inline src/generator.cpp:3303]: 4) This is a side effect of the main change where we now have to retrieve the initial value from `mInitialValueVariable` rather than `mVariable`.

File: tests/resources/deriv_approx_sin.xml
Comment (kerimoyle) [reply tests/resources/deriv_approx_sin.xml:6]: Yes we are ... that's still permitted, just to be noted in the tutorials and informative spec that it isn't recommended and will be removed soon.  If it's changed here, we need to check that we still have tests for initialisation from a variable until it's actually removed properly ...

**Comment by kerimoyle** [comment]
I've got some concerns about the "can't initialise a variable with another variable" thing.  We've just finished writing the spec that says that we do support this (at present) ... shouldn't we be maintaining tests and functionality in line with that?  I have no problem with this PR other than this sneaking idea that we might be missing stuff by making changes in the background ... ? @nickerso @hsorby @agarny Thoughts?

**Comment by nickerso** [comment]
> I've got some concerns about the "can't initialise a variable with another variable" thing

CellML allows this, the `develop` branch of libCellML does not yet support this for code generation. As code generation support for this improves this will be added back in. For this particular test I used the existing sine approximations model to save time, happy for the PR to be updated to use a different model to demonstrate this issue (which I think there were others that @agarny fixed as part of this so they could be sufficient).

File: tests/resources/deriv_approx_sin.xml
Comment (agarny) [reply tests/resources/deriv_approx_sin.xml:6]: I only meant that the code generator doesn't currently support the initialisation of a variable using another variable. This is clearly something that we need to add to the code generator. I have just created an issue for it (see issue #602).

Comment (kerimoyle) [reply tests/resources/deriv_approx_sin.xml:6]: Ah, ok, that makes more sense.  I thought we were removing functionality rather than just waiting for it :)

**Review by kerimoyle — APPROVED** [review]
OK with me.

**Comment by agarny** [comment]
> I've got some concerns about the "can't initialise a variable with another variable" thing. We've just finished writing the spec that says that we do support this (at present) ... shouldn't we be maintaining tests and functionality in line with that?

This needs to be done in another PR for which I have created an issue (see issue #602). Will work on that issue once we have merged in this PR and PR #599.

**Pull request merged**
